### PR TITLE
Custom plugin documentation

### PR DESCRIPTION
**Pull request by lornajane**
## What/Why/How?

Our custom plugins are brilliant, but the documentation is unclear and hard to follow. This pull request attempts to outline the available features, and add separate documentation for each thing, as well as clearer overviews on why each one is useful.

## Reference

https://redocly.com/docs/cli/resources/custom-plugins/

## Check yourself

- [ ] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [ ] All new/updated code is covered with tests

## Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### CODE REVIEW

**Comment by lornajane** [comment]
The various guides for decorators and custom rules need some updates (and some of this is in core now too, which is a great problem to have!), I'll do those separately since this is already plenty to review.

File: docs/custom-plugins/custom-config.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+## Configuration in plugins
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:1]:
```suggestion
# Configuration in plugins
```

File: docs/custom-plugins/custom-config.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Plugins can include configuration, under the `configs` key. This is
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:3]:
```suggestion
You can bundle rule configurations under the `configs` key in your plugin code. These configurations are
```

File: docs/custom-plugins/custom-config.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+useful when your plugin declares multiple rules or decorators, and you want to
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:4]:
```suggestion
useful when you declare multiple rules or decorators in your plugin, and you want to
```

File: docs/custom-plugins/custom-config.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+define within the plugin how to combine them.  Each plugin can have multiple
+bundles of configuration, and the user can use any of them in their
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:6]:
```suggestion
bundles of rule configurations, and the user can use any of them in their
```

File: docs/custom-plugins/custom-config.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+configuration file.
+
+
+Here's an example of configuration in a plugin, defining two bundles:
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:10]:
```suggestion
Here's an example plugin configuration, defining two rule bundles:
```

File: docs/custom-plugins/custom-config.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      rules: {
+        'operation-id-not-test': 'off',
+        'boolean-parameter-prefixes': 'error',
+      },
+    }
+  }
+};
+```
+
+With the example plugin above id as `my-local-plugin`, the `all` and `minimal`
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:32]:
```suggestion
In the example, the plugin id is `my-local-plugin` and the `all` and `minimal` rule
```

File: docs/custom-plugins/custom-config.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+configurations become available in the user's configuration file as
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:33]:
```suggestion
configurations are available in the user's configuration file as
```

File: docs/custom-plugins/custom-config.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+`"my-local-plugin/all"` and `"my-local-plugin/minimal"`, respectively. Use
+these configurations by adding them under `extends:` in `redocly.yaml`.
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:35]:
```suggestion
these configurations by adding them under `extends:` in your `redocly.yaml` configuration file.
```

File: docs/custom-plugins/custom-config.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+```yaml
+extends:
+  - my-local-plugin/all
+```
+
+Plugins can include rules, decorators and preprocessors in their configuration bundles.
```
Comment (HCloward) [inline docs/custom-plugins/custom-config.md:42]:
```suggestion
Plugins can include rules, decorators, and preprocessors in their configuration bundles.
```

File: docs/custom-plugins/custom-decorators.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
+# Decorators in plugins
+
+Decorators transform API descriptions, by adding, removing, or changing elements of the document. Before you build your own decorators:
+
+* [learn about Redocly decorators](../decorators.md)
+* [check the list of built-in decorators](../decorators.md#list-of-decorators)
+
+If you didn't find an existing decorator that fit your needs, then adding a decorator in a plugin is a good way to proceed.
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:8]:
```suggestion
If you can't find an existing decorator that fits your needs, then you can add a decorator in a plugin.
```

File: docs/custom-plugins/custom-decorators.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+:::warning Preprocessors
+
+Decorators and preprocessors are the same in structure, but preprocessors are run _before_ linting, and decorators are run after. We always recommend using decorators where possible, since the document might not be valid or structured as expected if the linting step hasn't run yet.
+
+:::
+
+## Plugin structure
+
+In order to create a preprocessor or decorator, the object that is exported from your module has to conform to the following interface:
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:18]:
```suggestion
To create a preprocessor or decorator, the object that is exported from your module has to conform to an interface such as the following example:
```

File: docs/custom-plugins/custom-decorators.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    oas3: {
+      "decorator-id": () => {
+        // ...
+      }
+    }
+  }
+}
+```
+
+Each decorator or preprocessor is a function that returns an object. The object's keys are the node types in the document, and each of those can contain any or all of the `enter()`, `leave()` and `skip()` functions for that node type. More information and examples on the [visitor pattern page](./visitor.md).
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:40]:
```suggestion
Each decorator or preprocessor is a function that returns an object. The object's keys are the node types in the document, and each of those can contain any or all of the `enter()`, `leave()` and `skip()` functions for that node type. Find more information and examples on the [visitor pattern page](./visitor.md).
```

File: docs/custom-plugins/custom-decorators.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        if(target.description) {
+          target.description = "✨ " + String(target.description);
+        }
+      }
+    },
+  }
+};
+```
+
+Decorators use the [visitor pattern](./visitor.md) to run an operation on every node in the document. In this example, when the code does `leave()` of an `Operation` node, it checks if the node (passed as `target` in this example) has a description, and updates it if it does.
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:65]:
```suggestion
Decorators use the [visitor pattern](./visitor.md) to run an operation on every node in the document. In this example, when the code executes the `leave()` function on the `Operation` node, it checks if the node (passed as `target` in this example) has a description, and updates it if it does.
```

File: docs/custom-plugins/custom-decorators.md [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+## Further examples of custom decorators
+
+See some more examples of decorators:
+- Follow our [replace-servers-url tutorial](../guides/replace-servers-url.md).
+- Change your [OAuth2 token URL](../guides/change-token-url.md).
+
+## Preprocessors
+
+Preprocessors follow the same structure and operation as decorators, but they are run before the validation/linting step. This makes them brittle because the document may not be valid, and the extra processing step can cause performance impacts. We recommend looking for alternative approaches to preprocessing.
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:156]:
```suggestion
Preprocessors follow the same structure and operation as decorators, but they are run before the validation/linting step. Running before the validation/linting step makes them brittle because the document may not be valid, and the extra processing step can cause performance impacts. We recommend looking for alternative approaches to preprocessing.
```

File: docs/custom-plugins/custom-decorators.md [hunk 15]
```diff
@@ -0,0 +1,158 @@
+# Decorators in plugins
+
+Decorators transform API descriptions, by adding, removing, or changing elements of the document. Before you build your own decorators:
+
+* [learn about Redocly decorators](../decorators.md)
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:5]:
```suggestion
* [Learn about Redocly decorators](../decorators.md)
```

File: docs/custom-plugins/custom-decorators.md [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+* [check the list of built-in decorators](../decorators.md#list-of-decorators)
```
Comment (HCloward) [inline docs/custom-plugins/custom-decorators.md:6]:
```suggestion
* [Check the list of built-in decorators](../decorators.md#list-of-decorators)
```

File: docs/custom-plugins/visitor.md [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+Here is a basic example of a nested visitor:
+
+```js
+function ExampleRule() {
+  const seen = {};
+  return {
+    Operation: {
+      // skip: (value, key) => ... // if needed
+      // enter(operation) {} // if needed
+      Schema(schema, ctx, parents) {
```
Comment (tatomyr) [inline docs/custom-plugins/visitor.md:64]: Wouldn't it be better to always explicitly use the `enter` visitor in our documentation? I remember, named visitor functions confused me when I saw them first time. It still looks strange to me that the same thing could be an object or a function  😅

Comment (lornajane) [reply docs/custom-plugins/visitor.md:64]: Ah, good spot. I didn't update this bit (just moved it to a new file) but I agree an object would be clearer. I'll switch the example.

Comment (lornajane) [reply docs/custom-plugins/visitor.md]: @tatomyr can you take another look? I tried to replace this existing example with an equivalent. It seems to work the same but would be good to get your eyes on it too

File: docs/custom-plugins/custom-rules.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  };
+}
+
+```
+
+The `ctx` object here holds all the context, which can be used to give more situation-aware functionality to the rules you build. This is one of the main use cases for custom rules. The `report()` method is used to give information to return to the user if the node being visited doesn't comply with the rule. You can read the [context](#the-context-object) and [location](#location-object) sections for more information.
+
+Adding this as part of a plugin requires you to add it to the `rules` part of the plugin object, under the relevant document type. The example rule here is intended to be used with OpenAPI, so the plugin code in `plugins/my-rules.js` is as follows:
+
+```
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:39]:
```suggestion
```js
```

File: docs/custom-plugins/custom-rules.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+The context object contains additional functionality that is helpful for rules to do their jobs. As the name implies, the context object contains information that is relevant to the context of the rule. The context object has the following properties:
+
+- `location` - current location in the source document. See [Location Object](#location-object)
+- `parentLocations` - mapping of parent node to its location (only for nested visitors)
+- `type` - information about current type from type tree
+- `parent` - parent object or array
+- `key` - key in parent object or array
+- `oasVersion` specific OAS minor version of current document (can be `oas2`, `oas3` or `oas3_1`).
+
+The context object also offers some additional functionality that
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:77]: This is odd - perhaps this was added as a mistake?

File: docs/custom-plugins/custom-rules.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+Validate your OpenAPI document with `redocly lint openapi.yaml`. With this rule enabled as shown, any `operationId` fields that are "test" cause the validation step to emit a warning. You can use the example here as a basis for building your own rules, and many rules can be included in a single plugin.
+
+## Object references
+
+### The context object
+
+The context object contains additional functionality that is helpful for rules to do their jobs. As the name implies, the context object contains information that is relevant to the context of the rule. The context object has the following properties:
+
+- `location` - current location in the source document. See [Location Object](#location-object)
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:70]:
```suggestion
- `location` - Current location in the source document. See [Location Object](#location-object).
```

File: docs/custom-plugins/custom-rules.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+- `parentLocations` - mapping of parent node to its location (only for nested visitors)
+- `type` - information about current type from type tree
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:72]:
```suggestion
- `type` - Information about the current type from the type tree.
```

File: docs/custom-plugins/custom-rules.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+- `parent` - parent object or array
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:73]:
```suggestion
- `parent` - Parent object or array.
```

File: docs/custom-plugins/custom-rules.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+- `key` - key in parent object or array
```
Comment (HCloward) [inline docs/custom-plugins/custom-rules.md:74]:
```suggestion
- `key` - Key in parent object or array.
```

File: docs/custom-plugins/extended-types.md [hunk 24]
```diff
@@ -0,0 +1,74 @@
+# Type extensions in plugins
+
+Redocly CLI in its core has a type tree which defines the structure of the API definition.
+Redocly CLI then uses it to do type-aware traversal of OpenAPI Document.
```
Comment (HCloward) [inline docs/custom-plugins/extended-types.md:4]:
```suggestion
Redocly CLI then uses it to do a type-aware traversal of an OpenAPI definition.
```

File: docs/custom-plugins/index.md [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+title: Custom plugins in Redocly
+redirectFrom:
+  - /docs/cli/custom-rules/
+  - /docs/cli/resources/custom-rules/
+  - /docs/cli/resources/custom-plugins/
+---
+
+# Custom plugins
+
+Custom plugins are a powerful way to extend Redocly. Use of custom plugins requires requires in-depth knowledge of the document type (such as OpenAPI) that you are working with, JavaScript coding skills, and the plugin interface.
```
Comment (HCloward) [inline docs/custom-plugins/index.md:11]:
```suggestion
Custom plugins are a powerful way to extend Redocly. Use of custom plugins requires in-depth knowledge of the document type (such as OpenAPI) that you are working with, JavaScript coding skills, and the plugin interface.
```

File: docs/custom-plugins/index.md [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+Custom plugins are recommended for advanced users who have exhausted the built-in options available, and who can develop JavaScript extensions themselves.
+Each plugin is a JavaScript module which can export custom rules, configuration, preprocessors, decorators or type tree extensions.
```
Comment (HCloward) [inline docs/custom-plugins/index.md:13]:
```suggestion
Each plugin is a JavaScript module which can export custom rules, configurations, preprocessors, decorators or type tree extensions.
```

File: docs/custom-plugins/index.md [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+
+:::Info
+For many users, the highly [configurable rules](../rules/configurable-rules.md) covers their needs without needing a custom plugin.
+:::
+
+Custom plugins use the [visitor pattern](./visitor.md) to traverse the document structure and apply the rules and decorators (and preprocessors if you have them) to each element. Rules defined in custom plugins may need additional context available, and this can be achieved using the [nested visitor approach](./visitor.md#nested-visitors) (please note that nested visitors are not available for decorators or preprocessors).
+
+## Extend Redocly with custom plugins
+
+The custom plugin functionality offers multiple ways to extend the built-in functionality of Redocly:
```
Comment (HCloward) [inline docs/custom-plugins/index.md:23]:
```suggestion
Custom plugins can extend the built-in functionality of Redocly in the following ways:
```

File: docs/custom-plugins/index.md [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+
+- **Write your own custom rules in JavaScript**. If your API standards include elements that can't be expressed by our existing rules, including configurable ones, you can code your own rules. See the [documentation on writing your own rules in custom plugins](./custom-rules.md).
```
Comment (HCloward) [inline docs/custom-plugins/index.md:25]:
```suggestion
- **Write your own custom rules in JavaScript**. If your API standards include elements that can't be expressed by our existing rules, including configurable ones, you can code your own rules. See the [documentation on writing your own rules in custom plugins](./custom-rules.md) for more information.
```

File: docs/custom-plugins/index.md [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+  - plugins/my-best-plugin.js
+  - plugins/another-plugin.js
+```
+
+The paths are relative to the configuration file location. Where there are multiple features that happen at the same time, such as decorators, the plugins loaded first in the `plugins` section are processed first.
+
+
+### Plugin structure
+
+The minimal plugin should export an `id` string that is used to refer to the contents of the plugin in configuration:
```
Comment (HCloward) [inline docs/custom-plugins/index.md:66]:
```suggestion
The minimal plugin should export an `id` string that is used to refer to the contents of the plugin in the `redocly.yaml` configuration file:
```

File: docs/custom-plugins/index.md [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+
+```js
+module.exports = {
+  id: 'my-local-plugin'
+}
+```
```
Comment (HCloward) [inline docs/custom-plugins/index.md:72]: Should we also include a sample `redocly.yaml` file with the id used?

File: docs/custom-plugins/index.md [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+
+<a id="oas-major-versions"></a>
+## Supported formats
+
+Everything that is exported from plugin relates to one of the supported document formats, such as OpenAPI v3. It is done by exporting an object containing a key-value mapping from a document format and version (`oas2` or `oas3` are supported) to an extension object (rules, preprocessors, decorators).
```
Comment (HCloward) [inline docs/custom-plugins/index.md:77]:
```suggestion
Everything that is exported from a plugin relates to one of the supported document formats, such as OpenAPI v3. Plugins work by exporting an object containing a key-value mapping from a document format and version (`oas2` or `oas3` are supported) to an extension object (rules, preprocessors, decorators).
```

File: docs/custom-plugins/index.md [hunk 32]
```diff
@@ -0,0 +1,81 @@
...
+
+Before processing the definition document Redocly CLI detects the document format and applies corresponding set of extensions.
```
Comment (HCloward) [inline docs/custom-plugins/index.md:79]:
```suggestion
Before processing the definition document, Redocly CLI detects the document format and applies a corresponding set of extensions.
```

File: docs/custom-plugins/visitor.md [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+To understand how this applies to your API description, think of the document as a tree structure. The top level elements are entries like `info` and `components`. To examine the `description` field in the `info` section, the visitor goes to the `info` node first, then on to the `description` node. This pattern is repeated all over the document as the visitor pattern works its way around the document tree.
+
+<a id="visitor-object"></a><a id="format-of-visitor-object"></a>
+
+## Structure of the visitor object
+
+In your plugin, create a JavaScript visitor object, and describe the functionality required for each type of node.
+
+Redocly CLI calls `enter()` while going down the tree and `leave()` going up the tree after processing the node and its children.
+`skip()` predicate is called and if it returns `true` the node is ignored for this visitor.
```
Comment (HCloward) [inline docs/custom-plugins/visitor.md:14]: This statement is confusing to me. When is the skip predicate called?

File: docs/custom-plugins/visitor.md [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+
+Top level **visitor functions** run only once for each node.
+If the same node is referenced by the $ref multiple times,
+top-level **visitor functions** are executed only once for this node.
+
+This works fine for most context-free rules which check basic things. If you need contextual info you should use [nested visitors](#nested-visitors).
+
+## Nested visitors
+
+Visitor object (if it is not `any` or `ref`) can define [nested visitors](#nested-visitors).
```
Comment (HCloward) [inline docs/custom-plugins/visitor.md:52]:
```suggestion
The visitor object (if it is not `any` or `ref`) can define [nested visitors](#nested-visitors).
```

File: docs/custom-plugins/visitor.md [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  };
+}
+```
+
+The `Schema` **visitor function** is called by Redocly CLI only if the Schema Object is encountered while traversing a tree while the Operation Object is **entered**.
+
+As the third argument, `enter()` in a **nested visitor object** accepts the `parents` object with corresponding parent nodes as defined in the **visitor object**.
+
+<div class="attention"> It will be executed only for the first level of Schema Object.</div>
```
Comment (HCloward) [inline docs/custom-plugins/visitor.md:74]:
```suggestion
<div class="attention"> It will be executed only for the first level of the Schema Object.</div>
```

File: docs/decorators.md [hunk 36]
```diff
@@ -4,22 +4,27 @@ redirectFrom:
 ---
 # Decorators
 
-All Redocly CLI decorators are listed on this page.
+Decorators are a way of changing an API description during bundling. This can be useful for many use cases:
```
Comment (HCloward) [inline docs/decorators.md:7]:
```suggestion
Decorators are a way of changing an API description during bundling. This updating during bundling can be useful for the following use cases:
```

File: docs/custom-plugins/custom-rules.md (hunk 19 above)
Comment (lornajane) [reply docs/custom-plugins/custom-rules.md:77]: Apparently my brain stalled mid-sentence? Thanks for catching this

File: docs/custom-plugins/index.md (hunk 26 above)
Comment (HCloward) [reply docs/custom-plugins/index.md:13]: So it might be an American/English thing, because from what I can tell configuration can be made plural with an "s" in American English and it is fine. I do not have particularly strong feelings either way though, so your choice.

File: docs/custom-plugins/index.md (hunk 30 above)
Comment (lornajane) [reply docs/custom-plugins/index.md:72]: I'm not sure it's that useful because this section is about how to structure the plugin, and how it should be used is different for each type of thing you put in your plugin, so I put the examples in each specific page.

File: docs/custom-plugins/visitor.md (hunk 33 above)
Comment (lornajane) [reply docs/custom-plugins/visitor.md:14]: Let me have another go. We visit each node in turn and call whichever of these methods is defined.

File: docs/custom-plugins/index.md (hunk 26 above)
Comment (lornajane) [reply docs/custom-plugins/index.md:13]: If it  seems off to you, it probably seems wrong to other US English speakers so I'm accepting the suggestion. Happy to hear that either would be valid though!

File: docs/custom-plugins/custom-config.md [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+extends:
+  - my-local-plugin/all
+```
+
+Plugins can include rules, decorators, and preprocessors in their configuration bundles.
+
+:::info Rulesets
+
+If your configuration includes only Redocly built-in rules and decorators, try
+using a [ruleset](..//rules.md#rulesets) to achieve this (it's simpler than a
```
Comment (adamaltman) [inline docs/custom-plugins/custom-config.md:47]:
```suggestion
using a [ruleset](../rules.md#rulesets) to achieve this (it's simpler than a
```
?

File: docs/custom-plugins/custom-decorators.md [hunk 38]
```diff
@@ -0,0 +1,158 @@
...
+        if(target.description) {
+          target.description = "✨ " + String(target.description);
+        }
+      }
+    },
+  }
+};
+```
+
+Decorators use the [visitor pattern](./visitor.md) to run an operation on every node in the document. In this example, when the code executes the `leave()` function on the `Operation` node, it checks if the node (passed as `target` in this example) has a description, and updates it if it does.
```
Comment (adamaltman) [inline docs/custom-plugins/custom-decorators.md:65]: question: Do you have to use `target` or can it be replaced with any string, such as `op`? I'm not clear for example if this also works:

```js
module.exports = OperationSparkle;

function OperationSparkle() {
  console.log("adding sparkles ... ");
  return {
    Operation: {
      leave(op) {
        if(op.description) {
          op.description = "✨ " + String(op.description);
        }
      }
    },
  }
};
```

File: docs/custom-plugins/custom-rules.md [hunk 39]
```diff
@@ -0,0 +1,117 @@
...
+- `parent` - Parent object or array.
+- `key` - Key in the parent object or array.
+- `oasVersion` - Specific OAS minor version of the current document (can be `oas2`, `oas3` or `oas3_1`).
+
+The context object also offers some additional functionality to resolve references and to return information about a problem to the user. The methods available are as follows:
+
+- `resolve(node)` - Synchronously dereferences `$ref` node to its value. Works only with `$refs` from the original document. If you need to resolve a reference from another source, you can use the optional second parameter: `resolve(node, from: string)`.
+- `report(descriptor)` - Reports a problem in the definition and returns information to the user. See [Report rule context](#report-rule-context) for more information.
+
+<a id="context-report"></a>
```
Comment (adamaltman) [inline docs/custom-plugins/custom-rules.md:82]: suggestion: Remove this anchor because this is a new page and there are no pre-existing links.
```suggestion
```

File: docs/custom-plugins/extended-types.md [hunk 40]
```diff
@@ -0,0 +1,74 @@
+# Type extensions in plugins
+
+Redocly CLI in its core has a type tree which defines the structure of the API definition.
+Redocly CLI then uses it to do a type-aware traversal of an OpenAPI definition.
```
Comment (adamaltman) [inline docs/custom-plugins/extended-types.md:4]: question: Is the plan to update this from saying "OpenAPI" once this merges? https://github.com/Redocly/redocly-cli/pull/1127

File: docs/custom-plugins/extended-types.md [hunk 41]
```diff
@@ -0,0 +1,74 @@
...
+const XMetaData = {
+  properties: {
+    lifecycle: { type: 'string'},
+    'owner-team': { type: 'string'},
+  },
+  required: ['lifecycle']
+};
+```
+
+Note the quotes around the `owner-team` key since it contains a hyphen `-`. These fields are strings but you can also define other datastructures if you need to represent something more complex. The [built-in type definitions](https://github.com/Redocly/redocly-cli/tree/main/packages/core/src) are a good place to look for examples.
```
Comment (adamaltman) [inline docs/custom-plugins/extended-types.md:36]:
```suggestion
Note the quotes around the `owner-team` key since it contains a hyphen `-`. These fields are strings but you can also define other data structures if you need to represent something more complex. The [built-in type definitions](https://github.com/Redocly/redocly-cli/tree/main/packages/core/src) are a good place to look for examples.
```

File: docs/custom-plugins/extended-types.md [hunk 42]
```diff
@@ -0,0 +1,74 @@
...
+    lifecycle: production
+    owner-team: Engineering/Integrations
+```
+
+Define the type extension for the new `x-metadata` section, in this example, we make the `lifecycle` field required.
+
+```js
+const XMetaData = {
+  properties: {
+    lifecycle: { type: 'string'},
```
Comment (adamaltman) [inline docs/custom-plugins/extended-types.md:29]:
```suggestion
    lifecycle: { type: 'string', enum: ['development', 'staging', 'production']},
```
suggestion: make it a little more than a basic string

File: docs/custom-plugins/index.md [hunk 43]
```diff
@@ -0,0 +1,81 @@
...
+
+The minimal plugin should export an `id` string that is used to refer to the contents of the plugin in the `redocly.yaml` configuration file:
+
+```js
+module.exports = {
+  id: 'my-local-plugin'
+}
+```
+
+<a id="oas-major-versions"></a>
```
Comment (adamaltman) [inline docs/custom-plugins/index.md:74]:
```suggestion
```
Suggestion: remove anchor because there are no links to this document or anchor that I could find. Maybe I'm  searching the wrong place?

File: docs/custom-plugins/visitor.md [hunk 44]
```diff
@@ -0,0 +1,111 @@
+# Visitor pattern
+
+_Visitor_ is a design pattern that allows operations to be performed on individual elements in a more complex structure. Redocly uses it as the basis for custom plugins where every node in a document must be evaluated, for example when applying rules or decorators.
+
+To understand how this applies to your API description, think of the document as a tree structure. The top level elements are entries like `info` and `components`. To examine the `description` field in the `info` section, the visitor goes to the `info` node first, then on to the `description` node. This pattern is repeated all over the document as the visitor pattern works its way around the document tree.
+
+<a id="visitor-object"></a><a id="format-of-visitor-object"></a>
```
Comment (adamaltman) [inline docs/custom-plugins/visitor.md:7]:
```suggestion
```
Suggestion: Remove -- I didn't find any anchor links to either of these anchors.

File: docs/custom-plugins/visitor.md [hunk 45]
```diff
@@ -0,0 +1,111 @@
...
+
+## Structure of the visitor object
+
+In your plugin, create a JavaScript visitor object, and describe the functionality required for each type of node.
+
+Redocly CLI calls `enter()` while going down the tree and `leave()` going up the tree after processing the node and its children.
```
Comment (adamaltman) [inline docs/custom-plugins/visitor.md:13]: idea: I'm usually not for diagrams or illustrations, because they can be hard to maintain. I could see a diagram adding some value here though. 

```
Root enter() --> Info enter() --> License enter() --> ... (turn around) ...  ↓
    --> License leave() --> Info leave() --> Root leave()
```

I don't think this is required to do (I mean to leave it as an idea for the future if we find people need more help with the concept).

File: docs/custom-plugins/visitor.md [hunk 46]
```diff
@@ -0,0 +1,111 @@
...
+If the `skip()` predicate is defined and returns `true` the node is ignored for this visitor.
```
Comment (adamaltman) [inline docs/custom-plugins/visitor.md:14]: question: Does this result in all children nodes being ignored too? I'm trying to understand the reason to skip something versus omitting it from the plugin.

File: docs/custom-plugins/visitor.md [hunk 47]
```diff
@@ -0,0 +1,111 @@
...
+- `ref` - visitor is called on $ref nodes.
+
+
+## Visitors execution and $ref
+
+Top level **visitor functions** run only once for each node.
+If the same node is referenced by the $ref multiple times,
+top-level **visitor functions** are executed only once for this node.
+
+This works fine for most context-free rules which check basic things. If you need contextual info you should use [nested visitors](#nested-visitors).
```
Comment (adamaltman) [inline docs/custom-plugins/visitor.md:48]:
```suggestion
This works fine for most context-free rules. If you need contextual info you should use [nested visitors](#nested-visitors).
```
suggestion: Not sure what is the line between basic and not basic? It seems like contextual info is the line, so maybe it is more clear with that removed because the next sentence addresses the context.

File: docs/custom-plugins/visitor.md [hunk 48]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  };
+}
+```
+
+The `Schema` **visitor function** is called by Redocly CLI only if the Schema Object is encountered while traversing a tree while the Operation Object is **entered**.
+
+As the third argument, `enter()` in a **nested visitor object** accepts the `parents` object with corresponding parent nodes as defined in the **visitor object**.
+
+<div class="attention"> It will be executed only for the first level of the Schema Object.</div>
```
Comment (adamaltman) [inline docs/custom-plugins/visitor.md:74]:
```suggestion
:::attention
It is executed only for the first level of the Schema Object.
:::
```

**Review by adamaltman — APPROVED** [review]
Great 🎉 

I left some suggestions, questions, and ideas.

File: docs/custom-plugins/custom-rules.md [hunk 49]
```diff
@@ -0,0 +1,117 @@
...
+Each rule is a function that returns an object with methods that Redocly CLI calls to "visit" nodes while traversing the definition document. The object keys are the node types that are encountered in the document. In this simple example, the custom plugin holds a rule that fails if any `operationId` is set to "test".
+
+To keep the plugin code manageable, each rule can go in its own file. This example is in `plugins/rules/opid-not-test.js`:
+
+```js
+module.exports = OperationIdNotTest;
+
+function OperationIdNotTest() {
+  return {
+    Operation(operation, ctx) {
```
Comment (tatomyr) [inline docs/custom-plugins/custom-rules.md:22]: One more place with implicit usage of  the visitor func.

**Review by tatomyr — APPROVED** [review]
A last comment. Overall looks good.

File: docs/custom-plugins/custom-decorators.md (hunk 38 above)
Comment (lornajane) [reply docs/custom-plugins/custom-decorators.md:65]: yes, I tried to use "target" in an attempt to make it a bit more copy/paste friendly, but called out that it was just an example, so I think it's clear.

File: docs/custom-plugins/extended-types.md (hunk 40 above)
Comment (lornajane) [reply docs/custom-plugins/extended-types.md:4]: I think we'll need to update a few things, yes! I intentionally made things more vague in a few places but at this time, I think the type extension feature is really only useful for OpenAPI.

File: docs/custom-plugins/custom-rules.md (hunk 49 above)
Comment (lornajane) [reply docs/custom-plugins/custom-rules.md]: Thanks! Fixed it :)
